migrate/20241203_authorization_refactor.rb

```diff
@@ -0,0 +1,135 @@
+        ["ffffffff-ff00-835a-87ff-f05a007343a0", "Postgres:edit", "ttzzzzzzzz021gzzz0pg0ed1t0"],
+        ["ffffffff-ff00-835a-87ff-f05a00d85dc0", "Postgres:view", "ttzzzzzzzz021gzzz0pg0v1ew1"],
+        ["ffffffff-ff00-835a-87c1-6801fc7343a0", "Postgres:fwedit", "ttzzzzzzzz021gz0pg0fwed1t1"],
+        ["ffffffff-ff00-835a-87c1-6801fcd85dc0", "Postgres:fwview", "ttzzzzzzzz021gz0pg0fwv1ew0"],
+        ["ffffffff-ff00-835a-87ff-ff8359029ad0", "SubjectTag:add", "ttzzzzzzzz021gzzzz0ts0add1"],
+        ["ffffffff-ff00-835a-87c1-ac830ea036e0", "SubjectTag:remove", "ttzzzzzzzz021gz0ts0rem0ve0"],
+        ["ffffffff-ff00-835a-87ff-f06b20d85dc0", "SubjectTag:view", "ttzzzzzzzz021gzzz0ts0v1ew1"],
+        ["ffffffff-ff00-835a-87ff-ff834a029ad0", "ActionTag:add", "ttzzzzzzzz021gzzzz0ta0add0"],
+        ["ffffffff-ff00-835a-87c1-a5030ea036e0", "ActionTag:remove", "ttzzzzzzzz021gz0ta0rem0ve1"],
+        ["ffffffff-ff00-835a-87ff-f06940d85dc0", "ActionTag:view", "ttzzzzzzzz021gzzz0ta0v1ew0"],
+        ["ffffffff-ff00-835a-87ff-ff8340029ad0", "ObjectTag:add", "ttzzzzzzzz021gzzzz0t00add0"],
+        ["ffffffff-ff00-835a-87c1-a0030ea036e0", "ObjectTag:remove", "ttzzzzzzzz021gz0t00rem0ve1"],
+        ["ffffffff-ff00-835a-87ff-f06800d85dc0", "ObjectTag:view", "ttzzzzzzzz021gzzz0t00v1ew0"]]
+            .each(&:pop))
+
+    %w[subject action object].each do |tag_type|
+      create_table(:"#{tag_type}_tag") do
+        uuid :id, primary_key: true
+        foreign_key :project_id, :project, type: :uuid, null: (tag_type == "action")
+        String :name, null: false
+
+        index [:project_id, :name], unique: true
+      end
+
+      create_table(:"applied_#{tag_type}_tag") do
+        foreign_key :tag_id, :"#{tag_type}_tag", type: :uuid
+        tag_column = :"#{tag_type}_id"
+        uuid tag_column
+
+        primary_key [:tag_id, tag_column]
+        index [tag_column, :tag_id]
+        constraint :no_self_tag, Sequel.~(tag_id: tag_column)
+      end
+    end
+
+    # Action Tag UUID primary keys created via:
+    # UBID.generate_vanity_action_tag(name).to_uuid
+    # This shows the generated UBIDs, but those are removed before inserting.
+    # These non-project specific action tags will be managed by Ubicloud developers.
+    action_tags =
+      [["ffffffff-ff00-834a-87ff-ff82d2028210", "Project:all", "tazzzzzzzz021gzzzz0pj0a110"],
+        ["ffffffff-ff00-834a-87ff-ff8374028210", "Vm:all", "tazzzzzzzz021gzzzz0vm0a111"],
+        ["ffffffff-ff00-834a-87ff-ff82d9028210", "PrivateSubnet:all", "tazzzzzzzz021gzzzz0ps0a111"],
+        ["ffffffff-ff00-834a-87ff-ff81fc028210", "Firewall:all", "tazzzzzzzz021gzzzz0fw0a110"],
+        ["ffffffff-ff00-834a-87ff-ff802b028210", "LoadBalancer:all", "tazzzzzzzz021gzzzz01b0a111"],
+        ["ffffffff-ff00-834a-87ff-ff82d0028210", "Postgres:all", "tazzzzzzzz021gzzzz0pg0a111"],
+        ["ffffffff-ff00-834a-87ff-ff828ea2dd80", "Member", "tazzzzzzzz021gzzzz0member0"]].each(&:pop)
```

Comment:
We've `Admin` as the subject tag but `Member` as the action tag. It's little confusing for me. I believe we should also create `Member` as a subject tag and create it in the projects by default. Most users won't interact with our access control view or take the time to understand it. They just want to use it, which is why we introduced `admin` and `member` managed policies.

Reply:
> 1. When inviting a new user, we should set 'admin' or preferably 'member' as the default role, similar to our previous system. Inviting a new user without any access is not a common practice.

The UI allows inviting a new user into any subject tag the inviting user has access to.  That would work for this, assuming a Member subject tag is added).

> 2. We could add a 'Subject/Role' selector to the 'User Management' page. This way, most users can choose from predefined roles (admin, member) without having to visit the access control page.

The issue with this is that accounts can be in multiple subject tags.  This type of selector only works if a user can only be in a single subject tag.  I'm fairly sure the current support for this with managed policies is broken, because changing a user from Admin to Member does not remove admin access, it just adds a new policy with member access.

> 3. Creating a 'Member' role for every project might be excessive. Instead, we could create it only when a customer assigns it to a project user, mirroring our current behavior.

Automatically creating a subject tag if it doesn't exist requires dealing with race conditions.  Having a choice between the two, I would prefer always generating a Member subject tag, and treat it like the Admin subject tag (a single ACE for Member subject tag, Member action tag, on all objects), not user-editable.

Reply:
> The issue with this is that accounts can be in multiple subject tags. This type of selector only works if a user can only be in a single subject tag. I'm fairly sure the current support for this with managed policies is broken, because changing a user from Admin to Member does not remove admin access, it just adds a new policy with member access.

Actually, it should rewrites all managed policies and remove from the admin policy.
https://github.com/ubicloud/ubicloud/blob/ed2a77722924583ae1187eeb417f8c8169a75963/routes/project/user.rb#L119C1-L130C14

> Automatically creating a subject tag if it doesn't exist requires dealing with race conditions. Having a choice between the two, I would prefer always generating a Member subject tag, and treat it like the Admin subject tag (a single ACE for Member subject tag, Member action tag, on all objects), not user-editable.

When I looked into our database, we have 1.581 admin policies and 28 member policies. So, creating a member tag for every project could be excessive. Not sure should we have it at all or not.

Reply:
> > The issue with this is that accounts can be in multiple subject tags. This type of selector only works if a user can only be in a single subject tag. I'm fairly sure the current support for this with managed policies is broken, because changing a user from Admin to Member does not remove admin access, it just adds a new policy with member access.
> 
> Actually, it should rewrites all managed policies and remove from the admin policy. https://github.com/ubicloud/ubicloud/blob/ed2a77722924583ae1187eeb417f8c8169a75963/routes/project/user.rb#L119C1-L130C14

Apologies, that does work correctly.  It's the token code I wrote that doesn't handle the situation correctly (looks like due to the `unless tokens.empty?` part).

> > Automatically creating a subject tag if it doesn't exist requires dealing with race conditions. Having a choice between the two, I would prefer always generating a Member subject tag, and treat it like the Admin subject tag (a single ACE for Member subject tag, Member action tag, on all objects), not user-editable.
> 
> When I looked into our database, we have 1.581 admin policies and 28 member policies. So, creating a member tag for every project could be excessive. Not sure should we have it at all or not.

I recommend not having it.  It wouldn't save much time for users that want to restrict access to projects.

Reply:
OK, then it makes sense to add the Member role by default for projects, and an access control entry for the Member subject tag, Member action tag, and NULL object tag (members can take all allowed member actions on any object in the project).  That won't cause significant complexity if we treat the subject tag and access control entry just as if the user created them (different from the Admin access control entry, which is not modifiable by the user).  I'll work on that.

Note that the previous flow (before managed policies) was too complex, because it required writing a JSON policy to express what you want.  This gives users a UI to manage the security permissions, which is much easier on the users (especially with #2458).

Reply:
@ozgune @enescakir I pushed a commit that adds a Member subject tag by default for projects, and added a default access control entry that allows members of that tag access to the same permissions as the current Member access policy.  This allows users to invite users directly into the Member tag, without having to use the access control system directly.

Reply:
@jeremyevans What do you think about adding subject selector to the user list if it has zero or one subject tag?

https://github.com/ubicloud/ubicloud/pull/2476